crates/core/src/node/proximity_cache.rs

```diff
@@ -0,0 +1,575 @@
+use std::collections::{HashMap, HashSet};
+use std::sync::Arc;
+use std::time::{Duration, Instant};
+
+use dashmap::DashMap;
+use freenet_stdlib::prelude::{ContractInstanceId, ContractKey};
+use serde::{Deserialize, Serialize};
+use tokio::sync::RwLock;
+use tracing::{debug, info, trace};
+
+use super::PeerId;
+
+/// Proximity cache manager - tracks what contracts this node and its neighbors are caching
+pub struct ProximityCacheManager {
+    /// Contracts we are caching locally (u32 hashes for efficiency)
+    my_cache: Arc<RwLock<HashSet<u32>>>,
+
+    /// What we know about our neighbors' caches
+    /// PeerId -> Set of contract hashes they're caching
+    neighbor_caches: Arc<DashMap<PeerId, NeighborCache>>,
+
+    /// Statistics for monitoring
+    stats: Arc<RwLock<ProximityStats>>,
+
+    /// Last time we sent a batch announcement
+    last_batch_announce: Arc<RwLock<Instant>>,
+
+    /// Pending removals to be sent in the next batch announcement
+    pending_removals: Arc<RwLock<HashSet<u32>>>,
+}
+
+#[derive(Clone, Debug)]
+struct NeighborCache {
+    /// Contract hashes this neighbor is caching
+    contracts: HashSet<u32>,
+    /// Last time we received an update from this neighbor
+    last_update: Instant,
+}
+
+#[derive(Clone, Debug, Default)]
+pub struct ProximityStats {
+    pub cache_announces_sent: u64,
+    pub cache_announces_received: u64,
+    pub updates_via_proximity: u64,
+    pub updates_via_subscription: u64,
+    pub false_positive_forwards: u64,
+}
+
+/// Message types for proximity cache protocol
+#[derive(Debug, Clone, Serialize, Deserialize)]
+#[allow(clippy::enum_variant_names)]
+pub enum ProximityCacheMessage {
+    /// Announce contracts we're caching (immediate for additions, batched for removals)
+    CacheAnnounce {
+        /// Contracts we're now caching
+        added: Vec<u32>,
+        /// Contracts we're no longer caching
+        removed: Vec<u32>,
+    },
+    /// Request neighbor's cache state (for new connections)
+    CacheStateRequest,
+    /// Response with full cache state
+    CacheStateResponse { contracts: Vec<u32> },
+}
+
+impl ProximityCacheManager {
+    pub fn new() -> Self {
+        Self {
+            my_cache: Arc::new(RwLock::new(HashSet::new())),
+            neighbor_caches: Arc::new(DashMap::new()),
+            stats: Arc::new(RwLock::new(ProximityStats::default())),
+            last_batch_announce: Arc::new(RwLock::new(Instant::now())),
+            pending_removals: Arc::new(RwLock::new(HashSet::new())),
+        }
+    }
+
+    /// Generate a u32 hash from a ContractInstanceId
+    fn hash_contract(contract_id: &ContractInstanceId) -> u32 {
```

Comment:
@sanity @iduartgomez We’re truncating a 32-byte `ContractInstanceId` down to the first four bytes. That can create collisions, different contracts may share the same hash. Could we store the full key using a proper hash?

Reply:
@netsirius You're absolutely right - using only 4 bytes creates significant collision risk. With the birthday paradox, we'd see ~50% collision probability with only ~65k contracts. Would suggest using u64 instead of u32 - still fast for hashing/lookup while dramatically reducing collision probability.